odoo/service/server.py

```diff
@@ -936,17 +937,94 @@ def start(self):
-            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-            self.socket.setblocking(0)
-            self.socket.bind((self.interface, self.port))
-            self.socket.listen(8 * self.population)
 
-    def stop(self, graceful=True):
+            # reload
+            if self.graceful_reload and os.environ.get('ODOO_HTTP_SOCKET_FD'):
+                self.socket = socket.fromfd(int(os.environ['ODOO_HTTP_SOCKET_FD']), family, socket.SOCK_STREAM)
+
+            # systemd socket activation
+            elif os.environ.get('LISTEN_FDS') == '1' and os.environ.get('LISTEN_PID') == str(os.getpid()):
+                SD_LISTEN_FDS_START = 3
+                self.socket = socket.fromfd(SD_LISTEN_FDS_START, family, socket.SOCK_STREAM)
+
+            # default
+            else:
+                self.socket = socket.socket(family, socket.SOCK_STREAM)
+                self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+                self.socket.setblocking(0)
+                self.socket.bind((self.interface, self.port))
+                self.socket.listen(8 * self.population)
+
+    def stop_long_polling(self):
         if self.long_polling_pid is not None:
             # FIXME make longpolling process handle SIGTERM correctly
             self.worker_kill(self.long_polling_pid, signal.SIGKILL)
             self.long_polling_pid = None
+
+    def reload_gracefully(self):
+        _logger.info('Reloading server')
+        pid = os.fork()
+        if pid != 0:
+            # keep the http listening socket open during _reexec() to ensure uptime
+            http_socket_fileno = self.socket.fileno()
+            flags = fcntl.fcntl(http_socket_fileno, fcntl.F_GETFD)
+            fcntl.fcntl(http_socket_fileno, fcntl.F_SETFD, flags & ~fcntl.FD_CLOEXEC)
+            os.environ['ODOO_HTTP_SOCKET_FD'] = str(http_socket_fileno)
+            os.environ['ODOO_READY_SIGHUP_PID'] = str(pid)
+            _reexec()  # stops execution
+
+        # child process handles old server shutdown
+        _logger.info("Waiting for new server to start ...")
+        self.phoenix_hatched = False
+
+        def sighup_handler(sig, frame):
+            self.phoenix_hatched = True
+
+        signal.signal(signal.SIGHUP, sighup_handler)
+
+        reload_timeout = time.time() + 60
+        while not self.phoenix_hatched and time.time() < reload_timeout:
+            time.sleep(0.1)
+
+        if not self.phoenix_hatched:
+            _logger.error("Server couldn't reload (check the updated code), stopping gracefully")
+        else:
+            _logger.info("New server has started, stopping old server gracefully")
+
+        self.stop_long_polling()
+
+        super().stop()
+        worker_processes = {}
+        for pid in self.workers:
+            worker_processes[pid] = psutil.Process(pid)
+            self.worker_kill(pid, signal.SIGINT)
+
+        limit = time.time() + self.timeout
```

Comment:
Just to be sure: that means the cron timeout is not taken into account?

Reply:
I think it makes sense to consider the cron timeout since we are going in a direction where we define more and more asynchronous tasks through cron triggers.
For the purpose of the SaaS, it would be acceptable. However, if we want to strictly follow the time limits, we may need a slightly more complex approach, e.g. by using 2 lists with 2 timeouts. Here is an **untested** suggestion: https://gist.github.com/nim-odoo/94f5cb047bdc6d924818ad3d26a43ce4